File: typescript/src/http/transport.ts
/**
 * Official PIA SDK - Resilient HTTP Transport with Exponential Backoff & Telemetry
 */

import type { ResolvedPiaConfig } from "../config";
import {
  ApiError,
  AuthenticationError,
  NetworkError,
  ParseError,
  PermissionError,
  RateLimitError,
  TimeoutError,
} from "../errors";
import type { RateLimitInfo, RequestOptions } from "../types";

const RETRYABLE_STATUS_CODES = new Set([408, 429, 500, 502, 503, 504]);
const SDK_VERSION = "1.0.1";

export class HttpTransport {
  private lastRateLimitInfo: RateLimitInfo = {};

  constructor(private readonly config: ResolvedPiaConfig) {}

  /**
   * Returns the most recent rate limit headers telemetry recorded by the SDK.
   */
  public getRateLimitInfo(): RateLimitInfo {
    return { ...this.lastRateLimitInfo };
  }

  /**
   * Executes an HTTP request with automatic retry, jittered backoff, and timeout.
   */
  public async request<T>(
    endpoint: string,
    method: "GET" | "POST" | "PUT" | "DELETE" = "GET",
    body?: unknown,
    options?: RequestOptions
  ): Promise<T> {
    const url = `${this.config.baseUrl}${endpoint.startsWith("/") ? "" : "/"}${endpoint}`;
    const maxRetries = options?.maxRetries ?? this.config.maxRetries;
    const timeoutMs = options?.timeoutMs ?? this.config.timeoutMs;

    let attempt = 0;

    while (true) {
      attempt++;
      this.config.logger.debug(
        `HTTP ${method} ${endpoint} (Attempt ${attempt}/${maxRetries + 1})`
      );

      const controller = new AbortController();
      let isTimedOut = false;

      const timer = setTimeout(() => {
        isTimedOut = true;
        controller.abort();
      }, timeoutMs);

      // Chain caller's abort signal if provided
      if (options?.signal) {
        if (options.signal.aborted) {
          clearTimeout(timer);
          throw options.signal.reason;
        }
        options.signal.addEventListener(
          "abort",
          () => {
            clearTimeout(timer);
            controller.abort(options.signal?.reason);
          },
          { once: true }
        );
      }

      try {
        const headers: Record<string, string> = {
          "x-api-key": this.config.apiKey,
          Accept: "application/json",
          "User-Agent": `pia-sdk-ts/${SDK_VERSION}`,
          ...this.config.headers,
          ...(options?.headers || {}),
        };

        let serializedBody: string | undefined;
        if (body !== undefined) {
          headers["Content-Type"] = "application/json";
          serializedBody = JSON.stringify(body);
        }

        const response = await this.config.fetch(url, {
          method,
          headers,
          body: serializedBody,
          signal: controller.signal,
        });

        clearTimeout(timer);

        // Extract rate limit and quota headers
        this.parseTelemetryHeaders(response.headers);

        if (response.ok) {
          return await this.parseResponseBody<T>(response, endpoint, method);
        }

        // Handle error responses
        const status = response.status;
        const errorText = await response.text().catch(() => "");
        let errorJson: any = null;
        try {
          if (errorText) errorJson = JSON.parse(errorText);
        } catch {
          // Ignored if not JSON
        }

        const errorMessage =
          errorJson?.message ||
          errorJson?.error ||
          (errorText ? errorText.slice(0, 200) : `HTTP ${status}`);

        const isRetryable = RETRYABLE_STATUS_CODES.has(status) && attempt <= maxRetries;

        if (isRetryable) {
          const delay = this.calculateBackoffDelay(attempt, response.headers);
          this.config.logger.warn(
            `Request to ${endpoint} failed with HTTP ${status}. Retrying in ${delay}ms...`
          );
          await this.sleep(delay);
          continue;
        }

        // Map non-retryable or exhausted errors
        this.handleHttpError(status, errorMessage, endpoint, method, response.headers, errorJson);
      } catch (err: unknown) {
        clearTimeout(timer);

        if (isTimedOut) {
          throw new TimeoutError(
            `Request to ${endpoint} exceeded timeout of ${timeoutMs}ms.`,
            timeoutMs,
            { endpoint, method }
          );
        }

        // If caller explicitly aborted
        if (options?.signal?.aborted) {
          throw options.signal.reason;
        }

        // If it's already a typed PiaError, rethrow
        if (err instanceof Error && "isPiaError" in err) {
          throw err;
        }

        // Check if transient network failure is retryable
        if (attempt <= maxRetries) {
          const delay = this.calculateBackoffDelay(attempt);
          this.config.logger.warn(
            `Network error calling ${endpoint}: ${err instanceof Error ? err.message : String(err)}. Retrying in ${delay}ms...`
          );
          await this.sleep(delay);
          continue;
        }

        throw new NetworkError(
          `Network transport failed connecting to ${endpoint}: ${err instanceof Error ? err.message : String(err)}`,
          err,
          { endpoint, method }
        );
      }
    }
  }

  private parseTelemetryHeaders(headers: Headers): void {
    const parseNum = (name: string): number | undefined => {
      const val = headers.get(name);
      if (!val) return undefined;
      const num = parseInt(val, 10);
      return Number.isFinite(num) ? num : undefined;
    };

    this.lastRateLimitInfo = {
      limit: parseNum("x-ratelimit-limit"),
      remaining: parseNum("x-ratelimit-remaining"),
      resetSeconds: parseNum("x-ratelimit-reset"),
      dailyLimit: parseNum("x-dailyquota-limit"),
      dailyRemaining: parseNum("x-dailyquota-remaining"),
    };
  }

  private calculateBackoffDelay(attempt: number, headers?: Headers): number {
    // Respect Retry-After header if provided
    const retryAfter = headers?.get("retry-after");
    if (retryAfter) {
      const parsedSeconds = parseInt(retryAfter, 10);
      if (Number.isFinite(parsedSeconds) && parsedSeconds > 0) {
        return Math.min(parsedSeconds * 1000, this.config.maxRetryDelayMs);
      }
    }

    // Exponential backoff with Full Jitter
    const base = this.config.retryDelayMs * Math.pow(2, attempt - 1);
    const capped = Math.min(base, this.config.maxRetryDelayMs);
    const jitter = 0.5 + Math.random() * 0.5;
    return Math.round(capped * jitter);
  }

  private async parseResponseBody<T>(
    response: Response,
    endpoint: string,
    method: string
  ): Promise<T> {
    const text = await response.text();
    if (!text || text.trim() === "") {
      return {} as T;
    }

    try {
      return JSON.parse(text) as T;
    } catch (err) {
      throw new ParseError(
        `Failed to parse JSON response from ${endpoint}: ${err instanceof Error ? err.message : String(err)}`,
        text.slice(0, 500),
        { statusCode: response.status, endpoint, method }
      );
    }
  }

  private handleHttpError(
    status: number,
    message: string,
    endpoint: string,
    method: string,
    headers: Headers,
    body: unknown
  ): never {
    const details = {
      statusCode: status,
      endpoint,
      method,
      requestId: headers.get("x-request-id") || headers.get("cf-ray"),
      rawResponse: body,
    };

    if (status === 401) {
      throw new AuthenticationError(message, details);
    }

    if (status === 403) {
      throw new PermissionError(message, undefined, details);
    }

    if (status === 429) {
      const retryAfter = headers.get("retry-after");
      const retryAfterSeconds = retryAfter ? parseInt(retryAfter, 10) : undefined;
      throw new RateLimitError(message, {
        retryAfterSeconds: Number.isFinite(retryAfterSeconds) ? retryAfterSeconds : undefined,
        dailyLimit: this.lastRateLimitInfo.dailyLimit,
        dailyRemaining: this.lastRateLimitInfo.dailyRemaining,
        minuteLimit: this.lastRateLimitInfo.limit,
        minuteRemaining: this.lastRateLimitInfo.remaining,
        details,
      });
    }

    throw new ApiError(message, status, body, details);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: python/pia/transport.py
"""Official PIA SDK - Resilient HTTP Transports (Sync & Async) with Exponential Backoff."""

from __future__ import annotations

import asyncio
import json
import logging
import random
import time
from typing import Any, Dict, Optional, Set

import httpx

from .config import PiaConfig
from .errors import (
    ApiError,
    AuthenticationError,
    NetworkError,
    ParseError,
    PermissionError,
    RateLimitError,
    TimeoutError,
)
from .logger import setup_logger
from .types import RateLimitInfo

SDK_VERSION = "1.0.1"
RETRYABLE_STATUS_CODES: Set[int] = {408, 429, 500, 502, 503, 504}


def _parse_telemetry(headers: httpx.Headers) -> RateLimitInfo:
    def _get_int(key: str) -> Optional[int]:
        val = headers.get(key)
        if val is not None:
            try:
                return int(val)
            except ValueError:
                return None
        return None

    return RateLimitInfo(
        limit=_get_int("x-ratelimit-limit"),
        remaining=_get_int("x-ratelimit-remaining"),
        reset_seconds=_get_int("x-ratelimit-reset"),
        daily_limit=_get_int("x-dailyquota-limit"),
        daily_remaining=_get_int("x-dailyquota-remaining"),
    )


def _calc_backoff(attempt: int, retry_after: Optional[str], config: PiaConfig) -> float:
    if retry_after:
        try:
            val = float(retry_after)
            if val > 0:
                return min(val, config.max_retry_delay)
        except ValueError:
            pass

    base = config.retry_delay * (2 ** (attempt - 1))
    capped = min(base, config.max_retry_delay)
    jitter = 0.5 + random.random() * 0.5
    return capped * jitter


def _handle_error_response(
    response: httpx.Response,
    endpoint: str,
    method: str,
    telemetry: RateLimitInfo,
) -> None:
    status = response.status_code
    req_id = response.headers.get("x-request-id") or response.headers.get("cf-ray")

    body_json: Any = None
    try:
        body_json = response.json()
    except Exception:
        pass

    error_msg = ""
    if isinstance(body_json, dict):
        error_msg = body_json.get("message") or body_json.get("error") or ""
    if not error_msg:
        error_msg = response.text[:200] if response.text else f"HTTP {status}"

    details = {
        "status_code": status,
        "endpoint": endpoint,
        "method": method,
        "request_id": req_id,
        "raw_response": body_json or response.text,
    }

    if status == 401:
        raise AuthenticationError(error_msg, **details)
    if status == 403:
        raise PermissionError(error_msg, **details)
    if status == 429:
        retry_val: Optional[int] = None
        raw_retry = response.headers.get("retry-after")
        if raw_retry:
            try:
                retry_val = int(raw_retry)
            except ValueError:
                pass
        raise RateLimitError(
            error_msg,
            retry_after_seconds=retry_val,
            daily_limit=telemetry.daily_limit,
            daily_remaining=telemetry.daily_remaining,
            minute_limit=telemetry.limit,
            minute_remaining=telemetry.remaining,
            **details,
        )

    raise ApiError(error_msg, **details)


class SyncTransport:
    """Synchronous resilient HTTP transport with exponential backoff and telemetry tracking."""

    def __init__(
        self,
        config: PiaConfig,
        client: Optional[httpx.Client] = None,
        logger: Optional[logging.Logger] = None,
    ) -> None:
        self.config = config
        self.logger = logger or setup_logger(config.debug)
        self._last_telemetry = RateLimitInfo()
        self.client = client or httpx.Client(
            base_url=config.base_url,
            timeout=config.timeout,
            headers={
                "x-api-key": config.api_key,
                "Accept": "application/json",
                "User-Agent": f"pia-sdk-py/{SDK_VERSION}",
                **config.headers,
            },
        )
        self._owns_client = client is None

    def get_rate_limit_info(self) -> RateLimitInfo:
        return self._last_telemetry

    def close(self) -> None:
        if self._owns_client:
            self.client.close()

    def request(
        self,
        endpoint: str,
        method: str = "GET",
        json_data: Any = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> Any:
        attempt = 0
        max_retries = self.config.max_retries

        url = f"{self.config.base_url}{endpoint if endpoint.startswith('/') else '/' + endpoint}"
        req_headers = {
            "x-api-key": self.config.api_key,
            "Accept": "application/json",
            "User-Agent": f"pia-sdk-py/{SDK_VERSION}",
            **self.config.headers,
            **(headers or {}),
        }

        while True:
            attempt += 1
            self.logger.debug("HTTP %s %s (Attempt %d/%d)", method, endpoint, attempt, max_retries + 1)

            try:
                response = self.client.request(
                    method=method,
                    url=url,
                    json=json_data,
                    params=params,
                    headers=req_headers,
                    timeout=timeout or self.config.timeout,
                )

                self._last_telemetry = _parse_telemetry(response.headers)

                if response.is_success:
                    if not response.text or not response.text.strip():
                        return {}
                    try:
                        return response.json()
                    except json.JSONDecodeError as err:
                        raise ParseError(
                            f"Failed to parse JSON response: {err}",
                            raw_text=response.text[:500],
                            status_code=response.status_code,
                            endpoint=endpoint,
                            method=method,
                        ) from err

                is_retryable = response.status_code in RETRYABLE_STATUS_CODES and attempt <= max_retries
                if is_retryable:
                    delay = _calc_backoff(attempt, response.headers.get("retry-after"), self.config)
                    self.logger.warning(
                        "Request to %s failed with HTTP %d. Retrying in %.2fs...",
                        endpoint,
                        response.status_code,
                        delay,
                    )
                    time.sleep(delay)
                    continue

                _handle_error_response(response, endpoint, method, self._last_telemetry)

            except httpx.TimeoutException as err:
                if attempt <= max_retries:
                    delay = _calc_backoff(attempt, None, self.config)
                    self.logger.warning("Request timed out on %s. Retrying in %.2fs...", endpoint, delay)
                    time.sleep(delay)
                    continue
                raise TimeoutError(
                    f"Request to {endpoint} timed out after {timeout or self.config.timeout}s.",
                    timeout_seconds=timeout or self.config.timeout,
                    endpoint=endpoint,
                    method=method,
                ) from err

            except httpx.NetworkError as err:
                if attempt <= max_retries:
                    delay = _calc_backoff(attempt, None, self.config)
                    self.logger.warning("Network error on %s: %s. Retrying in %.2fs...", endpoint, err, delay)
                    time.sleep(delay)
                    continue
                raise NetworkError(
                    f"Network error connecting to {endpoint}: {err}",
                    cause=err,
                    endpoint=endpoint,
                    method=method,
                ) from err


class AsyncTransport:
    """Asynchronous resilient HTTP transport for asyncio applications."""

    def __init__(
        self,
        config: PiaConfig,
        client: Optional[httpx.AsyncClient] = None,
        logger: Optional[logging.Logger] = None,
    ) -> None:
        self.config = config
        self.logger = logger or setup_logger(config.debug)
        self._last_telemetry = RateLimitInfo()
        self.client = client or httpx.AsyncClient(
            base_url=config.base_url,
            timeout=config.timeout,
            headers={
                "x-api-key": config.api_key,
                "Accept": "application/json",
                "User-Agent": f"pia-sdk-py/{SDK_VERSION}",
                **config.headers,
            },
        )
        self._owns_client = client is None

    def get_rate_limit_info(self) -> RateLimitInfo:
        return self._last_telemetry

    async def aclose(self) -> None:
        if self._owns_client:
            await self.client.aclose()

    async def request(
        self,
        endpoint: str,
        method: str = "GET",
        json_data: Any = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> Any:
        attempt = 0
        max_retries = self.config.max_retries

        url = f"{self.config.base_url}{endpoint if endpoint.startswith('/') else '/' + endpoint}"
        req_headers = {
            "x-api-key": self.config.api_key,
            "Accept": "application/json",
            "User-Agent": f"pia-sdk-py/{SDK_VERSION}",
            **self.config.headers,
            **(headers or {}),
        }

        while True:
            attempt += 1
            self.logger.debug("Async HTTP %s %s (Attempt %d/%d)", method, endpoint, attempt, max_retries + 1)

            try:
                response = await self.client.request(
                    method=method,
                    url=url,
                    json=json_data,
                    params=params,
                    headers=req_headers,
                    timeout=timeout or self.config.timeout,
                )

                self._last_telemetry = _parse_telemetry(response.headers)

                if response.is_success:
                    if not response.text or not response.text.strip():
                        return {}
                    try:
                        return response.json()
                    except json.JSONDecodeError as err:
                        raise ParseError(
                            f"Failed to parse JSON response: {err}",
                            raw_text=response.text[:500],
                            status_code=response.status_code,
                            endpoint=endpoint,
                            method=method,
                        ) from err

                is_retryable = response.status_code in RETRYABLE_STATUS_CODES and attempt <= max_retries
                if is_retryable:
                    delay = _calc_backoff(attempt, response.headers.get("retry-after"), self.config)
                    self.logger.warning(
                        "Request to %s failed with HTTP %d. Retrying in %.2fs...",
                        endpoint,
                        response.status_code,
                        delay,
                    )
                    await asyncio.sleep(delay)
                    continue

                _handle_error_response(response, endpoint, method, self._last_telemetry)

            except httpx.TimeoutException as err:
                if attempt <= max_retries:
                    delay = _calc_backoff(attempt, None, self.config)
                    self.logger.warning("Request timed out on %s. Retrying in %.2fs...", endpoint, delay)
                    await asyncio.sleep(delay)
                    continue
                raise TimeoutError(
                    f"Request to {endpoint} timed out after {timeout or self.config.timeout}s.",
                    timeout_seconds=timeout or self.config.timeout,
                    endpoint=endpoint,
                    method=method,
                ) from err

            except httpx.NetworkError as err:
                if attempt <= max_retries:
                    delay = _calc_backoff(attempt, None, self.config)
                    self.logger.warning("Network error on %s: %s. Retrying in %.2fs...", endpoint, err, delay)
                    await asyncio.sleep(delay)
                    continue
                raise NetworkError(
                    f"Network error connecting to {endpoint}: {err}",
                    cause=err,
                    endpoint=endpoint,
                    method=method,
                ) from err
